src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,113 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+
+    private static final double GROW_INDEX = 1.5;
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] data;
+    private int size;
+
+    public ArrayList() {
+        data = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkFreeSpace();
+        data[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
 
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add element: " + value
+                    + " on the position: " + index);
+        }
+        checkFreeSpace();
+        System.arraycopy(data, index, data, index + 1, size - index);
+        data[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index : " + index + " is not valid.");
+        }
+        return (T) data[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't set value. Index "
+                    + index + " doesn't exist.");
+        }
+        data[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't remove element from position: "
+                    + index);
+        }
+        return makeCopyAndDelete(index);
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (data[i] == element || data[i] != null && data[i].equals(element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Can't remove non-existing element: " + element);
+    }
+
+    private void checkFreeSpace() {
+        if (size == data.length) {
+            data = grow();
+        }
+    }
+
+    private Object[] grow() {
+        int newCapacity = (int) (data.length * GROW_INDEX);
+        return data = Arrays.copyOf(data, newCapacity);
```

Comment:
```suggestion
        int newCapacity = data.length * (data.length >> 1);
        Object[] destination = new Object[newCapacity];
        System.arraycopy(data, 0, destination,0, size);
        data = destination;
```
Make this method with void. Don't use Arrays class for copying data. You should use System.arraycopy() (It is pointed in checklist)